engines/e_dasync.c

```diff
@@ -15,6 +15,7 @@
  */
 #include "internal/deprecated.h"
 
+#include <openssl/opensslconf.h>
```

Comment:
Change that to include `openssl/macros.h`

Reply:
Or wait, hold on...  there's that inclusion of "internal/deprecated.h"...  this file shouldn't need all the guarding, so I wonder what goes wrong with it.  @mattcaswell might be able to answer that, maybe?

Reply:
Ah, never mind!  libcrypto.so doesn't export those symbols, so `e_dasync.c` can try to declare itself "internal" all it wants, that still won't work.

``` diff
diff --git a/engines/e_dasync.c b/engines/e_dasync.c
index d3536cf745..66207a27b5 100644
--- a/engines/e_dasync.c
+++ b/engines/e_dasync.c
@@ -7,15 +7,6 @@
  * https://www.openssl.org/source/license.html
  */
 
-/*
- * SHA-1 low level APIs are deprecated for public use, but still ok for
- * internal use.  Note, that due to symbols not being exported, only the
- * #defines and strucures can be accessed, in this case SHA_CBLOCK and
- * sizeof(SHA_CTX).
- */
-#include "internal/deprecated.h"
-
-#include <openssl/opensslconf.h>
 #if defined(_WIN32)
 # include <windows.h>
 #endif
@@ -23,6 +14,7 @@
 #include <stdio.h>
 #include <string.h>
 
+#include <openssl/macros.h>
 #include <openssl/engine.h>
 #include <openssl/sha.h>
 #include <openssl/aes.h>
```

Reply:
Additionally:

``` diff
diff --git a/engines/e_dasync.c b/engines/e_dasync.c
index d3536cf745..785b55d937 100644
--- a/engines/e_dasync.c
+++ b/engines/e_dasync.c
@@ -103,6 +95,8 @@ static int dasync_digest_nids(const int **nids)
 
 /* RSA */
 
+#ifndef OPENSSL_NO_DEPRECATED_3_0
+
 static int dasync_pub_enc(int flen, const unsigned char *from,
                     unsigned char *to, RSA *rsa, int padding);
 static int dasync_pub_dec(int flen, const unsigned char *from,
@@ -119,6 +113,8 @@ static int dasync_rsa_finish(RSA *rsa);
 
 static RSA_METHOD *dasync_rsa_method = NULL;
 
+#endif
+
 /* AES */
 
 static int dasync_aes128_cbc_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg,
```

Reply:
The comment details the two things that are needed from the `sha.h` header, neither requires exported symbols: one is a macro, the other the size of a structure.

Reply:
"internal/deprecated.h" does much more than that.  It turns off everything about deprecation (so you don't get the `OPENSSL_NO_DEPRECATED_3_0` macro, for example) and assumes that it has access to all symbols that are kept internally (such as the deprecated RSA symbols).  Unfortunately, since it links with libcrypto.so, you end up with lots of linking errors.

Reply:
Because of the `RSA_meth_` functions.

Reply:
Rather than fixing something that's going to be deprecated, I stopped building this engine.